Export plugins

mbasaglia opened the pull request:
Adds more metadata to the plugins to support a proper export system.

I've also updated the lottie exporter to use this system.

This is said plugin description as an example:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<plugin type="exporter">
   <_name>Export to Lottie format</_name>
   <exec>lottie-exporter.py</exec>
   <extension>.json</extension>
   <description>Lottie Animation (*.json)</description>
</plugin> 
```

if the `type` attribute is present and its value is "exporter", the plugin will be added to a list of exporters. The `<extension>` and `<description>` elements are used by the new "Export" dialog to set up the file filters.

Such plugins no longer show in the Plugins menu, they are only accessible through File > Export...

AnishGG commented:
@mbasaglia seems like great work! Did you test this locally? The travis build is failing with some errors.
Also as @morevnaproject proposed, we need a behaviour like this: https://github.com/synfig/synfig/issues/1218 where a file's undo history should not be erased. ^_^

AnishGG commented:
I tried building your PR locally and it worked. Don't know why is it failing on travis.
@ice0 can you have a look here. ^__^

AnishGG commented on synfig-studio/plugins/lottie-exporter/common/Param.py:
```diff
@@ -433,6 +433,7 @@ def recur_animate(self, anim_type):
                 return ret, self.expression_controllers
 
             elif self.param[0].tag == "bone_link":
+                import pdb; pdb.set_trace(); pass
```
why do we need debugger here?

AnishGG commented on synfig-studio/plugins/lottie-exporter/common/Param.py:
```diff
@@ -753,6 +754,7 @@ def __get_value(self, frame):
                     ret = link_on * switch + link_off * (1 - switch)
 
             elif self.param[0].tag == "bone_link":
+                import pdb; pdb.set_trace(); pass
```
same here

AnishGG commented on synfig-studio/plugins/lottie-exporter/common/Param.py:
```diff
@@ -893,6 +895,7 @@ def update_frame_window(self, window):
                 self.subparams["scale"].subparams["scalar"].update_frame_window(window)
 
             elif node.tag == "bone_link":
+                import pdb; pdb.set_trace(); pass
```
here as well

mbasaglia commented:
has sorry about pdb, I was using that when I was making my telegram exporter and forgot about it -_-

mbasaglia commented:
> Also as @morevnaproject proposed, we need a behaviour like this: #1218 where a file's undo history should not be erased. ^_^

It should already work on this branch, I've modified run_plugin with an argument that disables reloading the canvas.

I've also extracted the code that actually runs the plugin into a different function, which doesn't save a temp file, so it might be used in the future for things import plugins (that wouldn't require a canvas)

mbasaglia commented:
By the way, I was also considering adding support for a single plugin directory to define multiple plugins.

In the case of export, it might be helpful to define multiple exporters based on a single codebase.

Something like if the root element of the plugin xml file is `<plugins>`, it can define multiple `<plugin>` elements inside it.

mbasaglia commented:
Looks like the build errors on travis are about `Glib::RefPtr::get`, probably different Glib versions than what I have

AnishGG commented:
@ice0 could you review this please. ^__^

AnishGG commented:
Also @mbasaglia, as mentioned in #1218, could you change the menu item name from `export` to `export for web`.

mbasaglia commented:
> Also @mbasaglia, as mentioned in #1218, could you change the menu item name from `export` to `export for web`.

This is more general than just Lottie though, it can be used for any exporter. Main reason I made this is that I want to make my own exporters.

I already have a Telegram sticker exporter, and I might also add a SVG exporter.

mbasaglia commented:
I've added an extra commit just to make the XML parsing a bit simpler / more readable. I've also changed the `type` from an element into an attribute since that seems more appropriate.

rodolforg commented on synfig-studio/src/gui/instance.cpp:
```diff
@@ -223,25 +223,128 @@ Instance::set_redo_status(bool x)
+	if(custom_python_binary) {
+		command=custom_python_binary;
+		if (!App::check_python_version(command)) {
+			output=_("Error: You need to have Python 3 installed.");
```
This message is overriding by line 271 below. Besides, it doesn't make clear the searched python binary was from the env var.

rodolforg commented on synfig-studio/src/gui/instance.cpp:
```diff
@@ -223,25 +223,128 @@ Instance::set_redo_status(bool x)
+
+	bool result = false;
+	String output;
+	String command = "";
```
No need of this empty assignment :)

rodolforg commented on synfig-studio/src/gui/instance.cpp:
```diff
@@ -223,25 +223,128 @@ Instance::set_redo_status(bool x)
+		{
+			String python_path;
+#ifdef _WIN32
+			python_path = "\"" + App::get_base_path()+ETL_DIRECTORY_SEPARATOR+"python"+ETL_DIRECTORY_SEPARATOR+*iter+".exe" + "\"";
```
I know there's a lot of places that uses `ETL_DIRECTORY_SEPARATOR`, but I do believe it is better for readability to just use "/" instead: both Win, Mac & Linux accepts it as separator.

mbasaglia replied:
It's code that was already there, this kind of changes are beyond the scope of this merge request IMO.

rodolforg replied:
Ok.

rodolforg commented on synfig-studio/src/gui/instance.cpp:
```diff
@@ -223,25 +223,128 @@ Instance::set_redo_status(bool x)
+
+		FILE* pipe = popen(command.c_str(), "r");
+		if (!pipe) {
+			output = "ERROR: pipe failed!";
```
Hmm should it be translatable?

rodolforg commented on synfig-studio/src/gui/instance.cpp:
```diff
@@ -223,25 +223,128 @@ Instance::set_redo_status(bool x)
+        {
+            command += " \"" + arg + "\"";
+        }
+        command += " 2>&1";
```
Identation?

rodolforg commented on synfig-studio/src/gui/instance.cpp:
```diff
@@ -223,25 +223,128 @@ Instance::set_redo_status(bool x)
+			char buffer[128];
+			while(!feof(pipe)) {
+				if(fgets(buffer, 128, pipe) != NULL)
+						output += buffer;
```
Identation

rodolforg commented on synfig-studio/src/gui/instance.cpp:
```diff
@@ -223,25 +223,128 @@ Instance::set_redo_status(bool x)
 }
 
+bool
+studio::Instance::run_plugin_with_arguments(std::string plugin_path, const std::vector<std::string>& args)
```
I'd split up this function in two:
* One detects a valid python binary filepath - and maybe build the full command line
* Other runs the script (open pipe and everything)

mbasaglia replied:
I wanted to keep this merge request small so I haven't refactored much of the plugin code.

But I think it would also be a good idea to move most of the code in run_plugin into the plugin manager (or maybe even have plugin struct become a proper class and have it there).

`run_plugin_with_arguments` is really just extracting part of the code that used to be in `run_plugin`.

rodolforg commented on synfig-studio/src/gui/instance.cpp:
```diff
@@ -223,25 +223,128 @@ Instance::set_redo_status(bool x)
+                    this->get_canvas()->get_name().c_str());
 
-	String details = strprintf(_("This operation cannot be undone and all undo history will be cleared."));
+        String details = strprintf(_("This operation cannot be undone and all undo history will be cleared."));
```
You don't need to use `strprintf` if there isn't formatting placeholders.

rodolforg commented on synfig-studio/src/gui/instance.cpp:
```diff
@@ -296,93 +400,11 @@ studio::Instance::run_plugin(std::string plugin_path)
-			one_moment.show();
-
-		} else {
+		if (result && modify_canvas){
```
identation?

rodolforg commented on synfig-studio/src/gui/pluginmanager.cpp:
```diff
@@ -137,6 +137,8 @@ PluginManager::load_plugin( const std::string &path )
 	std::string plugindir = dirname(path);
 	p.id=plugindir;
+
+    std::list<plugin>* target_list = &list_;
```
identation

morevnaproject commented:
@mbasaglia Thank you for the great addition! I like how it work. ^__^

Just one request:
Currently the Export dialog allows to select file type (just one option at the moment - "Lottie JSON"). But when export is done it creates two files - "json" and "html". It would be nice to have two filetype options - "Lottie JSON" and "Lottie HTML". 

But that request can be a separate issue, I think this  PR is ready to merge. ^__^

mbasaglia commented:
> Just one request:
> Currently the Export dialog allows to select file type (just one option at the moment - "Lottie JSON"). But when export is done it creates two files - "json" and "html". It would be nice to have two filetype options - "Lottie JSON" and "Lottie HTML".

Yeah I'd like to add the ability to do something like that, maybe having some way of defining multiple plugins in the same directory.

It could be implemented by having multiple exporters in a single plugin but I think allowing multiple plugins in the same directory (sharing the same codebase) is more future-proof.

As I was also considering adding support for "import" plugins, so you could have a single installed plugin, that defines various functionality like both import and export in some format or the like.

The way inkscape handles this is by searching for files with a specific extension in the plugin directory (instead of just plugin.xml like synfig does). An alternative could be to add support different root element in plugin.xml that defines multiple `<plugin>` elements inside it.

I can work on this but I think the decision on how this is handled should be made by someone with more experience in this project.

morevnaproject commented:
@mbasaglia Can you please rebase your branch on top of the latest master, so I can make a clean merge? ^__^

mbasaglia commented:
@morevnaproject pushed rebased commits

morevnaproject commented:
@mbasaglia I have created new issue to continue discussion - https://github.com/synfig/synfig/issues/1309

rodolforg commented:
@ice0 @morevnaproject  maybe backport this and #1313 to v1.4.2?
